src/Microsoft.Identity.Web/TokenCacheProviders/TokenCacheExtensions.cs

```diff
@@ -81,20 +81,23 @@ public static class TokenCacheExtensions
                 throw new ArgumentNullException(nameof(initializeCaches));
             }
 
-            // Maintain a dictionary of service providers per `initializeCaches` delegate.
-            if (!s_serviceProviderFromAction.TryGetValue(initializeCaches.Method, out IServiceProvider? serviceProvider))
+            // try to reuse existing XYZ cache if AddXYZCache was called before, to simulate ASP.NET Core
+            var serviceProvider = s_serviceProviderFromAction.GetOrAdd(initializeCaches.Method, (m) =>
             {
-                IHostBuilder hostBuilder = Host.CreateDefaultBuilder()
-                    .ConfigureLogging(logger => { })
-                    .ConfigureServices(services =>
-                    {
-                        initializeCaches(services);
-                        services.AddDataProtection();
-                    });
+                lock (s_serviceProviderFromAction)
```

Comment:
After looking more closely at the fix for #1390 , I realize that there is still a race condition in between the 2 dictionary operations (which, by themselves, are thread safe).